File: app/backend/src/application/queries/handlers/get_product_handler.rs
use std::sync::Arc;

use crate::application::dto::ProductDTO;
use crate::application::error::ApplicationError;
use crate::application::queries::models::GetProductQuery;
use crate::application::repositories::ProductRepository;

/// 商品取得クエリハンドラ
/// CQRS パターンに基づく読み取り操作のハンドラ
pub struct GetProductHandler {
    product_repository: Arc<dyn ProductRepository + Send + Sync>,
}

impl GetProductHandler {
    pub fn new(product_repository: Arc<dyn ProductRepository + Send + Sync>) -> Self {
        Self { product_repository }
    }

    /// 商品取得クエリを実行
    ///
    /// # Arguments
    /// * `query` - 取得クエリデータ
    ///
    /// # Returns
    /// * `Result<ProductDTO, ApplicationError>` - 成功時は商品データ、失敗時はエラー
    pub async fn handle(&self, query: GetProductQuery) -> Result<ProductDTO, ApplicationError> {
        println!("->> get_product_handler: product_id={:?}", query.product_id);

        let product = self
            .product_repository
            .find_by_id(&query.product_id)
            .await?
            .ok_or(ApplicationError::ProductNotFound(
                query.product_id.to_string(),
            ))?;

        Ok(product)
    }
}

File: app/backend-ts/src/application/queries/handlers/get-product.handler.ts
import { Inject } from '@nestjs/common';
import { IQueryHandler, QueryHandler } from '@nestjs/cqrs';

import { ProductDto } from '$application/dto';
import {
  NotFoundError,
  ValidationError,
} from '$application/errors/application.error';
import { GetProductQuery } from '$application/queries';
import { IProductRepository } from '$application/repositories';
import { ProductId } from '$domain/value-objects';

@QueryHandler(GetProductQuery)
export class GetProductHandler implements IQueryHandler<GetProductQuery> {
  constructor(
    @Inject('IProductRepository')
    private readonly productRepository: IProductRepository,
  ) {}

  async execute(query: GetProductQuery): Promise<ProductDto> {
    // Parse and validate product ID
    let productId: ProductId;
    try {
      productId = ProductId.fromUuid(query.productId);
    } catch {
      throw new ValidationError(
        `Invalid product ID format: ${query.productId}`,
      );
    }

    // Fetch product DTO directly
    const productDto = await this.productRepository.findById(productId);
    if (!productDto) {
      throw new NotFoundError('Product', query.productId);
    }

    return productDto;
  }
}
